### DISCUSSION TITLE
Invalid/Valid memory ranges for `byte`/`half`/`word` accesses

### DISCUSSION THREAD

**Discussion by ghost** [psx-spx/psx-spx.github.io#28] [Ideas]
I've generated these IO maps by running a full address space scan. The maps themselves should be fairly straightforward, but the format is this:

```
% $min - $max | error=(0|1)
```

`error=0` means that the range _didn't_ cause an invalid address exception at each address in the range.
`error=1` means that the range _did_ cause an invalid address exception at each address in the range.

Some of these ranges are interesting, and I've never seen any documentation of them, for example:

```
% $fffe0000 - $fffe001f | error=0
% $fffe0100 - $fffe013f | error=0
```

`$fffe0130` is somewhat documented, but nothing else in that range is. I have no clue what any of these do, if they do anything at all.

[map-byte.md](https://github.com/psx-spx/psx-spx.github.io/files/7377387/map-byte.md)
[map-half.md](https://github.com/psx-spx/psx-spx.github.io/files/7377388/map-half.md)
[map-word.md](https://github.com/psx-spx/psx-spx.github.io/files/7377389/map-word.md)

The code to generate these maps is roughly (replace `read32` and a few `4`s for the other sizes):

```c
#include "exception.h"
#include "psx.h"

uint32_t exception_handler_80[4];

extern void as_exn_handler();

int main() {
  // Set up a custom exception handler that just sets $k1 to 1 and returns.

  save_exception_handler(0x80000080, (uint32_t *) exception_handler_80);
  load_exception_handler(0x80000080, (uint32_t *) as_exn_handler);

  // Start scanning the address space

  uint32_t address = 0;
  uint32_t address_of_last_change = 0;
  bool in_error_region = false;

  do {
    asm volatile("li $k1, 0");

    volatile uint32_t data = read32(address);
    bool error_happened;

    asm volatile("move %0, $k1" : "=r"(error_happened));

    if (in_error_region != error_happened) {
      bios_printf("$%08x - $%08x | error=%d\n",
        address_of_last_change,
        address - 4,
        in_error_region);

      address_of_last_change = address;
      in_error_region = error_happened;
    }
  } while (address += 4);

  // Restore previous exception handler

  load_exception_handler(0x80000080, (uint32_t *) exception_handler_80);

  return 0;
}
```
```asm
.set mips1
.set noreorder
.set noat

.text

.global as_exn_handler
.type as_exn_handler, function

as_exn_handler:
    j as_exn_handler_
    nop
    nop
    nop

as_exn_handler_:
    mfc0 $k0, $14
    li   $k1, 1
    nop
    addi $k0, 4
    jr   $k0
    rfe
```

**Comment by rixnobis** [comment]
Your scan matches what the page now documents. The unlocked ranges are listed under "Garbage Locations in I/O Area" in https://psx-spx.consoledev.net/unpredictablethings/, `FFFE0000h` (32 bytes) and `FFFE0100h` (48 bytes) included, with how reads behave there. The locked gaps around them are in the list just above. `FFFE0130h` is BCC. Closing.
